arrows/vxl/color_commonality_filter.cxx

```diff
@@ -0,0 +1,561 @@
+  auto const np = input.nplanes();
+  auto const istep = input.istep();
+  auto const jstep = input.jstep();
+  auto const pstep = input.planestep();
+
+  InputType const* row = input.top_left_ptr();
+  for( unsigned j = 0; j < nj; ++j, row += jstep )
+  {
+    InputType const* pixel = row;
+    for( unsigned i = 0; i < ni; ++i, pixel += istep )
+    {
+      InputType const* plane = pixel;
+      std::ptrdiff_t step = 0;
+      for( unsigned p = 0; p < np; ++p, plane += pstep )
+      {
+        step += histsteps[ p ] * ( *plane >> bitshift );
+      }
+      output( i, j ) = static_cast< OutputType >( *( hist_top_left + step ) );
+    }
+  }
+}
+
+// ----------------------------------------------------------------------------
+/// Create an output image indicating the relative commonality of each
+/// input pixel's color occuring in the entire input image. A lower
+/// value in the output image corresponds to that pixels value being
+/// less common in the entire input image.
+///
+/// Functions by first building a histogram of the input image, then,
+/// for each pixel, looking up the value in the histogram and scaling
+/// this value by a given factor.
+template < class InputType, class OutputType >
+void
+perform_filtering( vil_image_view< InputType > const& input,
+                   vil_image_view< OutputType >& output,
+                   color_commonality_filter_settings const& options )
+{
+  assert( std::numeric_limits< InputType >::is_integer );
+  assert( is_power_of_two( options.resolution_per_channel ) );
+
+  // Set output image size
+  output.set_size( input.ni(), input.nj() );
+
+  // If we are in grid mode, simply call this function recursively
+  if( options.grid_image )
+  {
+    color_commonality_filter_settings recursive_options = options;
+    recursive_options.grid_image = false;
+
+    // Formulate grided regions
+    unsigned ni = input.ni();
+    unsigned nj = input.nj();
+
+    for( unsigned j = 0; j < options.grid_resolution_height; ++j )
+    {
+      for( unsigned i = 0; i < options.grid_resolution_width; ++i )
+      {
+        // Top left point for region
+        int ti = ( i * ni ) / options.grid_resolution_width;
+        int tj = ( j * nj ) / options.grid_resolution_height;
+
+        // Bottom right
+        int bi = ( ( i + 1 ) * ni ) / options.grid_resolution_width;
+        int bj = ( ( j + 1 ) * nj ) / options.grid_resolution_height;
+
+        // Formulate rect region
+        vgl_box_2d< int > region{ ti, bi, tj, bj };
+
+        vil_image_view< InputType > region_data_ptr =
+          point_view_to_region( input, region );
+        vil_image_view< OutputType > output_data_ptr =
+          point_view_to_region( output, region );
+
+        // Process rect region independent of one another
+        perform_filtering( region_data_ptr, output_data_ptr,
+                           recursive_options );
+      }
+    }
+
+    // We are done, we already processed the entire image
+    return;
```

Comment:
It might make more sense to instead put the remainder of the function in an `else`?